servlet/src/main/java/com/linecorp/armeria/server/servlet/DefaultServletHttpRequest.java

```diff
@@ -0,0 +1,624 @@
+                if (!body.isEmpty()) {
+                    params2 = QueryParams.fromQueryString(body);
+                }
+            }
+
+            if (params1 == null || params1.isEmpty()) {
+                return firstNonNull(params2, QueryParams.of());
+            } else if (params2 == null || params2.isEmpty()) {
+                return params1;
+            } else {
+                return QueryParams.builder()
+                                  .sizeHint(params1.size() + params2.size())
+                                  .add(params1)
+                                  .add(params2)
+                                  .build();
+            }
+        } catch (Exception e) {
+            // If we failed to decode the query string, we ignore the exception raised here.
+            // A missing parameter might be checked when invoking the annotated method.
+            logger.debug("Failed to decode query string: {}", query, e);
+            return QueryParams.of();
+        }
+    }
+
+    /**
+     * Get aggregated http request.
+     */
+    @VisibleForTesting
+    AggregatedHttpRequest getHttpRequest() {
+        return httpRequest;
+    }
+
+    /**
+     * Parse cookie.
+     */
+    @Nullable
+    private Cookie[] decodeCookie() {
+        final String cookieValue = httpRequest.headers().get(HttpHeaderNames.COOKIE);
+        if (cookieValue != null) {
+            final Cookies cookies = com.linecorp.armeria.common.Cookie.fromCookieHeader(cookieValue);
+            return cookies.stream().map(c -> {
+                final Cookie cookie = new Cookie(c.name(), c.value());
+                if (c.domain() != null) {
+                    cookie.setDomain(c.domain());
+                }
+                if (c.path() != null) {
+                    cookie.setPath(c.path());
+                }
+                cookie.setSecure(c.isSecure());
+                cookie.setMaxAge(Ints.saturatedCast(c.maxAge()));
+                cookie.setHttpOnly(c.isHttpOnly());
+                return new Cookie(c.name(), c.value());
+            }).toArray(Cookie[]::new);
+        } else {
+            return null;
+        }
+    }
+
+    /**
+     * Parse path info.
+     */
+    private String decodePathInfo() {
+        final int index = getContextPath().length() + servletPath.length();
+        return index < requestURI.length() ? requestURI.substring(index) : null;
+    }
+
+    @Override
+    @Nullable
+    public Cookie[] getCookies() {
+        return cookies;
+    }
+
+    @Override
+    public long getDateHeader(String name) {
+        requireNonNull(name, "name");
+        for (DateFormat x : FORMATS_TEMPLATE) {
+            try {
+                final Date date = x.parse(getHeader(name));
+                if (date != null) {
+                    return date.getTime();
+                }
+            } catch (Exception e) {
+                logger.info("Try parse " + getHeader(name) + " to date");
```

Comment:
This can log three times for just one header.
How about logging it after all parsing fails?